File: Graph/***Topological Sort (DFS).cpp
Approach:

1. We can only have a topological sort only if the graph is DAG(Directed Acyclic).

#CODE:

void dfs(int node, vector<int> adj[],vector<int> &vis, stack<int> &st) {
    vis[node] = 1;
    for(auto it: adj[node]) {
        if(!vis[it]) {
            dfs(it,adj,vis,st);
        }
    }
    st.push(node);
}
vector<int> topoSort(int n, vector<int> adj[]) {
    vector<int> vis(n,0);
    stack<int> st;
    for(int i=0;i<n;++i) {
        if(!vis[i]) {
            dfs(i,adj,vis,st);
        }
    }
    vector<int> ans;
    while(!st.empty()) {
        ans.push_back(st.top());
        st.pop();
    }
    return ans;
}
=> Time Complexity = O(N + E)
=> Space Complexity = O(N)

File: Graph/***Topological Sort (DFS).java
Approach:

1. We can only have a topological sort only if the graph is DAG(Directed Acyclic).

#CODE:
class Solution
{   
    static void dfs(int node, ArrayList<ArrayList<Integer>> adj, int vis[], Stack<Integer> st) {
        vis[node] = 1;
        for(Integer it: adj.get(node)) {
            if(vis[it] == 0) {
                dfs(it, adj, vis, st);
            }
        }
        st.push(node);
    }
    static int[] topoSort(int V, ArrayList<ArrayList<Integer>> adj) 
    {
        int vis[] = new int[V];
        Arrays.fill(vis, 0);
        Stack<Integer> st = new Stack<Integer>();
        
        for(int i=0;i<V;++i) {
            if(vis[i] == 0) {
                dfs(i, adj, vis, st);
            }
        }
        
        int ans[] = new int[V];
        int idx = 0;
        while(!st.isEmpty()) {
            Integer node = st.pop();
            ans[idx++] = node;
        }
        return ans;
    }
}